File: 卡码/代码随想录/Go/回溯/01-组合/LeetCode_0216.go
//216. 组合总和 III
package main

func combinationSum3(k int, n int) [][]int {
	path:=make([]int,0)
	res:=make([][]int,0)
	var backtracking func(k,n,sum,index int)
	backtracking=func(k,n,sum,index int){
		if sum>n {
			return
		}
		if k==len(path) {
			if sum==n {
				tmp:=make([]int,k)
				copy(tmp,path)
				res=append(res,tmp)
			}
			return
		}
		for i:=index;i<=10-(k-len(path));i++ {
			path=append(path,i)
			backtracking(k,n,sum+i,i+1)
			path=path[:len(path)-1]
		}
	}
	backtracking(k,n,0,1)
	return res
}

File: 卡码/代码随想录/C++/回溯/01-组合/LeetCode_0216.cpp
//216. 组合总和 III
#include <iostream>
#include <vector>
using namespace std;

class Solution {
	vector<int> path;
	vector<vector<int>> res;
public:
	void backtracking(int k,int n,int sum,int index){
		if(sum>n) return;
		if(k==path.size()){
			if(sum==n) res.push_back(vector<int>(path));
			return;
		}
		for(int i=index;i<=10-(k-path.size());i++){
			path.push_back(i);
			backtracking(k,n,sum+i,i+1);
			path.pop_back();
		}
	}
	vector<vector<int>> combinationSum3(int k, int n) {
		backtracking(k,n,0,1);
		return res;
	}
};

int main(){
  system("pause");
  return 0;
}
